src/java_binding/src/lib.rs

```diff
@@ -816,6 +821,121 @@ pub extern "system" fn Java_com_risingwave_java_binding_Binding_rowClose<'a>(
     pointer.drop()
 }
 
+#[inline(never)]
+pub fn run_this_func_to_get_valid_ptr_from_java_binding() {}
+
+#[no_mangle]
+pub extern "system" fn Java_com_risingwave_java_binding_Binding_sendMsgToChannel<'a>(
+    mut env: EnvParam<'a>,
+    channel: Pointer<'a, MyJniSender>,
+    mut msg: JObject<'a>,
+) -> jboolean {
+    // If msg is null means just check whether channel is closed.
+    if msg.is_null() {
+        if channel.as_ref().is_closed() {
+            // Drop channel as well.
+            channel.drop();
+            return JNI_FALSE;
+        } else {
+            return JNI_TRUE;
+        }
+    }
+
+    let source_id = env
+        .env
+        .call_method(&mut msg, "getSourceId", "()J", &[])
+        .unwrap();
+    let source_id = source_id.j().unwrap();
+
+    let events_list = env
+        .env
+        .call_method(&mut msg, "getEventsList", "()Ljava/util/List;", &[])
+        .unwrap();
+    let mut events_list = match events_list {
+        JValueGen::Object(obj) => obj,
+        _ => unreachable!(),
+    };
+
+    let size = env
+        .env
+        .call_method(&mut events_list, "size", "()I", &[])
+        .unwrap()
+        .i()
+        .unwrap();
+    let mut events = Vec::with_capacity(size as usize);
+    for i in 0..size {
+        let java_element = env
+            .call_method(
```

Comment:
Extracting fields via `call_method` may be not a good practice. I guess serializing into bytes (in native buffer) on Java side and deserialize bytes on Rust side might be more efficient, and it can simply reuse the Protobuf message definition that we already have. 

Anyway, this is just an improvement and can be done later.